### PR TITLE
add single precision (float32) mode

### LINKED ISSUE #69: Query on halting single precision support
Hello,

I've been interested in using the single precision mode in `lleaves`. 

I noticed there was an initial implementation in the `vectorize` branch ([link](https://github.com/siboehm/lleaves/tree/vectorize)), but development seems to have halted. 

Could you share the reasons behind discontinuing this feature? 

Thanks for your great work on `lleaves`.

### PR DESCRIPTION
**Pull request by zjzjwang**
Closes #69 

mostly copy from this branch [vecterize](https://github.com/siboehm/lleaves/tree/vectorize)

### CODE REVIEW

File: tests/test_benchmark.py [hunk 1]
```diff
@@ -15,13 +15,13 @@ def test_benchmarks_correct_output():
     df = pd.read_parquet(
         "benchmarks/data/yellow_tripdata_2016-01.parquet", columns=NYC_used_columns
     )
-    NYC_X = feature_enginering().fit_transform(df).astype(np.float32)
+    NYC_X = feature_enginering().fit_transform(df).astype(np.float64)
 
     df = pd.read_csv("benchmarks/data/airline_data_factorized.csv")
-    airline_X = df.to_numpy(np.float32)
+    airline_X = df.to_numpy(np.float64)
```
Comment (siboehm) [inline tests/test_benchmark.py:21]: Huh I never noticed that I was accidetnally running these with float32

**Review by siboehm — CHANGES_REQUESTED** [review]
This is great stuff! Thank you for implementing it, changes look good apart from some minor interface & variable naming issues.

**Review by siboehm — CHANGES_REQUESTED** [review]
This is great stuff! Thank you for implementing it, changes look good apart from some minor interface & variable naming issues.

File: lleaves/lleaves.py [hunk 2]
```diff
@@ -42,15 +45,21 @@ class Model:
     # prediction function, drops GIL on entry
     _c_entry_func = None
 
-    def __init__(self, model_file):
+    def __init__(self, model_file, dtype="float64"):
```
[the remark this answers is not in the document]
Comment (zjzjwang) [reply lleaves/lleaves.py:48]: I originally thought the same, but seeing you put it in [`__init__`](https://github.com/siboehm/lleaves/blob/36677dde5116ddd9d4bf978b0506e7d8d7911a14/lleaves/lleaves.py#L46), I wrote it that way too.

File: lleaves/compiler/codegen/codegen.py [hunk 3]
```diff
@@ -33,6 +33,15 @@ def dconst(value):
     return ir.Constant(DOUBLE, value)
 
 
+def get_fdtype_const(value, double_precision):
+    f = dconst if double_precision else fconst
+    return f(value)
+
+
+def get_fdtype(double_precision):
```
[the remark this answers is not in the document]
Comment (zjzjwang) [reply lleaves/compiler/codegen/codegen.py]: Changed to `use_fp64`.

File: lleaves/compiler/tree_compiler.py [hunk 4]
```diff
@@ -9,6 +9,7 @@
 
 def compile_to_module(
     file_path,
+    double_precision=True,
```
[the remark this answers is not in the document]
Comment (zjzjwang) [reply lleaves/compiler/tree_compiler.py]: Sure.

**Comment by siboehm** [comment]
LGTM, thanks for the PR @zjzjwang! Will try to do a release on the weekend

**Pull request merged**